### PR TITLE
Graph Traversal: Handle non-collecting type-preserving edges

### PR DESCRIPTION
**Pull request by sachindshinde**
This PR ports `advancePathWithNonCollectingAndTypePreservingTransitions()` and its downstream methods.

There is one TODO introduced about populating dead-end information, but it's only necessary for composition (tracked in FED-173).

### CODE REVIEW

**Comment by goto-bus-stop** [comment]
I merged in the changes from #261 and #226 that have landed in main. Calling `build_query_plan` with this PR now yields:
```
Was not able to find any options for userById(id: 1) { name email }: This shouldn’t have happened.
```

File: src/query_graph/graph_path.rs [hunk 1]
```diff
@@ -1020,6 +1154,449 @@ where
...
+            Option<(Arc<GraphPath<TTrigger, TEdge>>, QueryPlanCost)>;
+        let mut best_path_by_source: IndexMap<NodeStr, BestPathInfo<TTrigger, TEdge>> =
+            IndexMap::new();
+        let dead_ends = vec![];
+        // Note that through `excluded` we avoid taking the same edge from multiple options. But
+        // that means it's important we try the smallest paths first. That is, if we could in theory
+        // have path A -> B and A -> C -> B, and we can do B -> D, then we want to keep A -> B -> D,
+        // not A -> C -> B -> D.
+        let mut heap: BinaryHeap<HeapElement<TTrigger, TEdge>> = BinaryHeap::new();
+        while let Some(HeapElement(to_advance)) = heap.pop() {
```
Comment (goto-bus-stop) [inline src/query_graph/graph_path.rs:1210]: `heap` at this point is empty, so the loop never runs.

The heap should be initialised with `self`. Can we get away with storing a reference to a path in the heap? If not, you can make this function only work on `Arc<Self>` like:

```rust
fn advance_with_non_collecting_and_type_preserving_transitions(
    self: &Arc<Self>,
    ...
) {
    ...
    heap.push(Arc::clone(self));
    ...
}
```

Then users can only call this method through an `Arc<GraphPath>` and the rest of the function Just Works.

Comment (sachindshinde) [reply src/query_graph/graph_path.rs]: Nice, the caller uses `Arc` so should be good!

File: src/query_graph/graph_path.rs [hunk 2]
```diff
@@ -401,7 +402,21 @@ pub(crate) struct SimultaneousPathsWithLazyIndirectPaths {
...
+impl ExcludedDestinations {
+    fn is_excluded(&self, destination: &NodeStr) -> bool {
+        self.0.contains(destination)
+    }
+
+    fn add_excluded(&self, destination: NodeStr) -> Self {
+        let mut new = self.clone();
+        if !self.is_excluded(&destination) {
+            Arc::make_mut(&mut new.0).push(destination)
+        }
```
[the remark this answers is not in the document]
Comment (sachindshinde) [reply src/query_graph/graph_path.rs]: Makes sense, thanks for the Rust info on `make_mut()`!

File: src/query_graph/graph_path.rs [hunk 3]
```diff
@@ -1020,6 +1154,449 @@ where
...
+
+impl<TTrigger, TEdge> PartialEq for HeapElement<TTrigger, TEdge>
+where
+    TTrigger: Eq + Hash,
+    Arc<TTrigger>: Into<GraphPathTrigger>,
+    TEdge: Copy + Into<Option<EdgeIndex>>,
+    EdgeIndex: Into<TEdge>,
+{
+    fn eq(&self, other: &HeapElement<TTrigger, TEdge>) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
```
[the remark this answers is not in the document]
Comment (sachindshinde) [reply src/query_graph/graph_path.rs]: So the behavior we want here is that `GraphPath`'s of different `Arc` pointers should be considered different/not equal, but we want a min heap ordered by path length; when `GraphPath`s are "equal", we don't really care which one we get back from `heap.pop()` as long as its path length isn't bigger than any other element in the heap.

My worry with letting length be equality was that pushing two elements with the same length would result in some kind of deduplication (I didn't see much in the BinaryHeap API docs about duplicates). I tried it out just now though with:
```rust
let mut heap = std::collections::BinaryHeap::new();
heap.push(1);
heap.push(1);
assert_eq!(heap.len(), 2)
```
and it works, so it seems like the current behavior of `BinaryHeap` is fine there. (And I'm guessing that behavior should be stable.)

[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/query_graph/graph_path.rs]: Right, the docs are a bit lacking here. Reading between the lines, I’d say when pushing elements that are distinguishable but compare equal it’s unspecified (but probably deterministic) which one comes out first, like an unstable sort.

**Pull request merged**